Resolve MakeGenericType ILLink warning in DependencyInjection

eerhardt opened the pull request:
Resolve the ILLink warning in DependencyInjection by adding a runtime check that is behind a new AppContext switch 'Microsoft.Extensions.DependencyInjection.VerifyOpenGenericServiceTrimmability'. The runtime check ensures the trimming annotations on the open generic types are compatible between the service and implementation types. The check is enabled by default when PublishTrimmed=true.

Here is a simple repro of the trimming issue where the developer doesn't see a warning:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <RuntimeIdentifier>win-x64</RuntimeIdentifier>
    <TrimmerDefaultAction>link</TrimmerDefaultAction>
    <PublishTrimmed>true</PublishTrimmed>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.Extensions.DependencyInjection" Version="6.0.0-*" />
  </ItemGroup>
</Project>
```

```C#
using Microsoft.Extensions.DependencyInjection;
using System;
using System.Diagnostics.CodeAnalysis;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            IServiceProvider sp = new ServiceCollection()
                .AddSingleton(typeof(IFactory<>), typeof(Factory<>))
                .BuildServiceProvider();

            var factory = sp.GetService<IFactory<Customer>>();

            Console.WriteLine(factory.Create());
        }
    }

    interface IFactory<T>
    {
        T Create();
    }

    class Factory<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicParameterlessConstructor)] T>
        : IFactory<T>
    {
        public T Create()
        {
            return Activator.CreateInstance<T>();
        }
    }

    class Customer { }
}
```

If you publish this app, it doesn't work:

```
Unhandled exception. System.MissingMethodException: Cannot dynamically create an instance of type 'ConsoleApp1.Customer'. Reason: No parameterless constructor defined.
   at System.RuntimeType.ActivatorCache..ctor(RuntimeType ) in System.Private.CoreLib.dll:token 0x600041c+0x93
   at System.RuntimeType.CreateInstanceOfT() in System.Private.CoreLib.dll:token 0x60003e7+0xf
   at System.Activator.CreateInstance[T]() in System.Private.CoreLib.dll:token 0x6000601+0x0
   at ConsoleApp1.Factory`1.Create() in C:\Users\eerhardt\source\repos\WorkerService4\ConsoleApp1\Program.cs:line 31
   at ConsoleApp1.Program.Main(String[] ) in C:\Users\eerhardt\source\repos\WorkerService4\ConsoleApp1\Program.cs:line 17
```

Today, the developer sees a warning coming from DependencyInjection:

```
warning IL2104: Assembly 'Microsoft.Extensions.DependencyInjection' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries
```

That is the warning being suppressed here and being validated by the new runtime check.

eerhardt commented:
Heads up @pranavkm - when this change gets to the aspnetcore repo, we will probably need to suppress some new warnings.

davidfowl commented:
This change doesn't look right. This is a type system limitation that will cause things to warn unnecessarily just in case you use a generic type in that slot. This should only warn for open generic types not closed ones. It's too aggressive

eerhardt commented:
> This should only warn for open generic types not closed ones.  It's too aggressive

Agreed. But given the current feature set of the trimmer, I don't see how it is possible to make it less aggressive.

cc @vitek-karas @marek-safar @agocke @MichalStrehovsky @sbomer

davidfowl commented:
Let it break for open generics and build an analyzer to detect open generics used in these methods. This is way too aggressive for my liking.

sbomer commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ServiceDescriptor.cs:
```diff
@@ -13,12 +13,15 @@ namespace Microsoft.Extensions.DependencyInjection
     [DebuggerDisplay("Lifetime = {Lifetime}, ServiceType = {ServiceType}, ImplementationType = {ImplementationType}")]
     public class ServiceDescriptor
     {
+        internal const string RequiresUnreferencedCode = "Using generic types in DependencyInjection can lead to required code being trimmed. Ensure any generic Service types (e.g. 'IMyService<T>') have matching trimming annotations as the Implementation types implementing them (e.g. 'MyService<[DynamicallyAccessedMembers] T>).";
```
```suggestion
        internal const string RequiresUnreferencedCode = "Using generic types in DependencyInjection can lead to required code being trimmed. Ensure any generic Service types (e.g. 'IMyService<T>') have the same trimming annotations as the Implementation types implementing them (e.g. 'MyService<[DynamicallyAccessedMembers] T>).";
```

sbomer commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/Extensions/ServiceCollectionDescriptorExtensions.cs:
```diff
@@ -129,6 +129,9 @@ public static class ServiceCollectionDescriptorExtensions
         /// <param name="collection">The <see cref="IServiceCollection"/>.</param>
         /// <param name="service">The type of the service to register.</param>
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "AddSingleton has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
```suggestion
            Justification = "ServiceDescriptor has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

sbomer commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/Extensions/ServiceCollectionDescriptorExtensions.cs:
```diff
@@ -217,6 +221,9 @@ public static class ServiceCollectionDescriptorExtensions
         /// <typeparam name="TService">The type of the service to add.</typeparam>
         /// <param name="collection">The <see cref="IServiceCollection"/>.</param>
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "AddSingleton has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
```suggestion
            Justification = "TryAddTransient has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

sbomer commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/Extensions/ServiceCollectionDescriptorExtensions.cs:
```diff
@@ -270,6 +278,9 @@ public static class ServiceCollectionDescriptorExtensions
         /// <param name="collection">The <see cref="IServiceCollection"/>.</param>
         /// <param name="service">The type of the service to register.</param>
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "AddSingleton has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
```suggestion
            Justification = "ServiceDescriptor has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

sbomer commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/Extensions/ServiceCollectionDescriptorExtensions.cs:
```diff
@@ -358,6 +370,9 @@ public static class ServiceCollectionDescriptorExtensions
         /// <typeparam name="TService">The type of the service to add.</typeparam>
         /// <param name="collection">The <see cref="IServiceCollection"/>.</param>
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "AddSingleton has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
```suggestion
            Justification = "TryAddScoped has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

sbomer commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/Extensions/ServiceCollectionDescriptorExtensions.cs:
```diff
@@ -411,6 +427,9 @@ public static class ServiceCollectionDescriptorExtensions
         /// <param name="collection">The <see cref="IServiceCollection"/>.</param>
         /// <param name="service">The type of the service to register.</param>
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "AddSingleton has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
```suggestion
            Justification = "ServiceDescriptor has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

sbomer commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/Extensions/ServiceCollectionDescriptorExtensions.cs:
```diff
@@ -499,6 +519,9 @@ public static class ServiceCollectionDescriptorExtensions
         /// <typeparam name="TService">The type of the service to add.</typeparam>
         /// <param name="collection">The <see cref="IServiceCollection"/>.</param>
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "AddSingleton has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
```suggestion
            Justification = "TryAddSingleton has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

sbomer commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ServiceCollectionServiceExtensions.cs:
```diff
@@ -107,6 +109,9 @@ public static class ServiceCollectionServiceExtensions
         /// <returns>A reference to this instance after the operation has completed.</returns>
         /// <seealso cref="ServiceLifetime.Transient"/>
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "AddSingleton has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
```suggestion
            Justification = "AddTransient has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

sbomer commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ServiceCollectionServiceExtensions.cs:
```diff
@@ -298,6 +305,9 @@ public static class ServiceCollectionServiceExtensions
         /// <returns>A reference to this instance after the operation has completed.</returns>
         /// <seealso cref="ServiceLifetime.Scoped"/>
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "AddSingleton has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
```suggestion
            Justification = "AddScoped has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

sbomer commented on src/libraries/Microsoft.Extensions.Hosting.Systemd/src/SystemdHostBuilderExtensions.cs:
```diff
@@ -40,11 +41,19 @@ public static IHostBuilder UseSystemd(this IHostBuilder hostBuilder)
         }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "ServiceDescriptor has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
```suggestion
            Justification = "AddSingleton has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

sbomer commented on src/libraries/Microsoft.Extensions.Logging.Configuration/src/LoggingBuilderConfigurationExtensions.cs:
```diff
@@ -14,6 +15,9 @@ public static class LoggingBuilderConfigurationExtensions
         /// </summary>
         /// <param name="builder">The <see cref="ILoggingBuilder"/> to register services on.</param>
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "ServiceDescriptor has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
```suggestion
            Justification = "TryAddSingleton has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

sbomer commented on src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs:
```diff
@@ -290,11 +290,17 @@ public static IHostBuilder UseConsoleLifetime(this IHostBuilder hostBuilder, Act
         }
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "ServiceDescriptor has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
```suggestion
            Justification = "AddSingleton has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

sbomer commented on src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:
```diff
@@ -282,5 +283,14 @@ private void CreateServiceProvider()
         }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "ServiceDescriptor has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
```suggestion
            Justification = "AddSingleton has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

sbomer commented on src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetimeHostBuilderExtensions.cs:
```diff
@@ -67,5 +67,13 @@ public static IHostBuilder UseWindowsService(this IHostBuilder hostBuilder, Acti
         }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "ServiceDescriptor has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
```suggestion
            Justification = "AddSingleton has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

sbomer commented on src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs:
```diff
@@ -19,6 +19,9 @@ public static class OptionsServiceCollectionExtensions
         /// <param name="services">The <see cref="IServiceCollection"/> to add the services to.</param>
         /// <returns>The <see cref="IServiceCollection"/> so that additional calls can be chained.</returns>
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "ServiceDescriptor's have RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
```suggestion
            Justification = "ServiceDescriptor has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

sbomer commented on src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientFactoryServiceCollectionExtensions.cs:
```diff
@@ -20,6 +20,9 @@ public static class HttpClientFactoryServiceCollectionExtensions
         /// <param name="services">The <see cref="IServiceCollection"/>.</param>
         /// <returns>The <see cref="IServiceCollection"/>.</returns>
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "ServiceDescriptor has RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```
This one looks like it silences warnings from ServiceDescriptor static methods, and also from TryAddTransient/TryAddSingleton - I wonder if that's worth mentioning in the justification.
```suggestion
            Justification = "ServiceDescriptor, TryAddTransient, and TryAddSingleton have RequiresUnreferencedCode because Service and Implementation types could be generic with mis-matching trimming annotations. " +
```

davidfowl commented:
I think this should be a forcing function to improve generics, attributes and the linker. We shouldn't make the library unusable because of those limitations.

MichalStrehovsky commented:
> Let it break for open generics and build an analyzer to detect open generics used in these methods. This is way too aggressive for my liking.

We do not break user code without warnings. Suppression of a warning that can result in broken code after trimming is a buggy suppression - not all dynamic patterns can be described statically - it's laws of physics. If this pattern ends up being general enough warranting a illinker feature to describe it, we can add an illinker feature to describe it.

davidfowl commented:
This can be described statically and I'd be ok if it warned for just open generics but we don't currently have a way to describe that. I don't think it's reasonable to make everything suffer because of that one scenario.

MichalStrehovsky commented:
> I don't think it's reasonable to make everything suffer because of that one scenario.

What's the failure mode when the constructor gets stripped? Will the user get actionable feedback on what to do next?

The warning added here states what the potential problem can be. I would not describe it as "suffering".

I've seen a lot of failure modes caused by tools stripping necessary things. That is a lot closer to what I see when someone says "suffering":

1. It happens after one already debugger their app and they're ready to ship (hitting Publish instead of F5 debug).
2. The failure mode is obscure (ranging from an exception saying something wasn't found, through a NullReferenceException to, "a control doesn't show up" or "nothing happens if I click a button").

davidfowl commented:
> What's the failure mode when the constructor gets stripped? Will the user get actionable feedback on what to do next?

The failure mode is based on what happens with the generic type. I don't know how they would know what to do next as its really dependent on the usage of the T in the concrete implementation of the generic.

jkotas commented:
Is it an option to add new set of APIs that disallow the open generic types and that are trimming friendly? Then the fix for this warning would be to use the new trimming friendly API, and the warnings would be only issues for the really problematic cases.

vitek-karas commented:
@eerhardt  and I discussed this offline and we both like the analyzer solution proposed by @davidfowl, just "improved".

The way I see this is that this is a pattern for which we don't have a solution in the trimmer. We don't even know what the solution would look like yet. And we won't get one in .NET 6 (too late for this I think).

I also think (based on my own perception as well as feedback from @davidfowl) that we can't accept the PR as is - it would produce too much noise. The assumption is that vast majority of the callsites for the problematic methods are actually completely trim compatible, we just can't detect it. So generating warnings for all of them is just super noisy.

My assumptions:
* We won't have a proper solution in 6 (be it trimmer based or new APIs)
* We think it's too important to not to anything about this (as in accept this PR)

So I accept that this is a problem we don't have a nice solution for in .NET 6 but at the same time we want at least some solution. So using a "hacky" solution seems like an OK thing to do - as long as we see it as a point-in-time change.

The "hacky" solution would be to build understanding of this into the linker:
* Linker would recognize these methods intrinsically (using custom step infra so that it's not hardwired into the core logic) and check that the types involved are open generic, and only then issue a warning. For all other cases it would not warn.
* We would build similar capability into our analyzer
* We would suppress the warning in the framework (the reason why this PR exists) - with a good comment and how to remove it later on when we do have a proper solution.

This is not too different from what we're doing for example for operators and `System.Linq.Expressions`... we're also patching a problem with intrinsic linker behavior.

Obviously I would like to avoid doing this, but if we think this is too important the above solution doesn't sound too bad.

jkotas commented:
> My assumptions:
> We think it's too important to not to anything about this 

It would be useful to clarify why it is must-have to do something about this for .NET 6.

eerhardt commented:
> why it is must-have to do something about this for .NET 6.

The current behavior isn't acceptable in my opinion. If I `dotnet publish` the repro project at the top of this PR, I _**do**_ get a warning:

```
warning IL2104: Assembly 'Microsoft.Extensions.DependencyInjection' produced trim warnings. For more information see https://aka.ms/dotnet-illink/libraries 
```

Navigating to that link brings me to a page that shows me how to make my own library trimmable - not what I can do because `Microsoft.Extensions.DependencyInjection` isn't fully trim-compatible.

As a user, I have no idea what is the problem with DependencyInjection. And this is the same warning I get for something that 100% doesn't work with trimming - like System.ComponentModel.Composition, System.CodeDom, or System.DirectoryServices. However, 95% of DependencyInjection works just fine with trimming. It is just this one last issue - open generic services - that is causing a problem. If there was a way to just flag those usages, I think the user experience would be perfect.

The situation would be slightly different if this was a one-off, rarely used library. But `Microsoft.Extensions.DependencyInjection.Abstractions` has been downloaded over one billion times - half a million a day. It is an extremely used library, and is the heart of all our other "Microsoft.Extensions" libraries. It is going to get used in even more places going forward with Maui apps taking a dependency on it.

For these reasons, I think this is something we must do in .NET 6 to complete our "make the .NET Libraries trim compatible" user story https://github.com/dotnet/runtime/issues/43078.

jkotas commented:
I agree that this is a popular library. But I do not think doing something about it in .NET 6 will move the needle significantly from the user scenario point of view. The app models where this library is used will generated a ton of other impossible to reason about warnings in .NET 6. (For example, I have just tried `dotnet new web + dotnet publish -r win-x64 /p:PublishTrimmed=true`.)

I understand that the special casing in the linker is always the easy way out. I do not think it is sustainable. If we want to do something about this, we should be looking at fixing this properly, and not by special casing in the linker.

eerhardt commented:
> The app models where this library is used will generated a ton of other impossible to reason about warnings in .NET 6.

Even Maui?

davidfowl commented:
We should absolutely try our best to make the trim warnings accurate and safe but this PR isn't the solution. I like the idea that the linker would have knowledge of this open generic "issue" but it would be nice if there was some attribute to indicate you wanted tho behavior on these types of methods.

jkotas commented:
> Even Maui?

I have not seen much done to make Maui trimming safe by construction. https://github.com/dotnet/maui/issues/355 that tracks the work does not have a lot of traffic. For .NET 6, I assume that Maui will be on trim and hope for the best plan that Xamarin apps have been on historically.

Correct me if my conjecture is wrong and we have plan in place to make maui trimming safe by construction for .NET 6.

We have done good work on making core libraries annotated for trimming in .NET 6. We have run out of time to address the trimming issues in the app models where I think M.E.DependencyInjection injection belongs. Another issue in the same bucket are issues with integration of source generated json serializers in ASP.NET that prevent ASP.NET webservices from being provably trimming safe by default. It will be something to focus on .NET 7. I expect that it will need work from multiple teams to address successfully (appmodel teams, runtime teams, language). 

> this PR isn't the solution.

This PR is the solutions with the tools we have at our disposal in .NET 6.

davidfowl commented:
> This PR is the solutions with the tools we have at our disposal in .NET 6.

Right but I'm not willing to sign off on it because I think its the wrong solution (even though it's limited by the tools at our disposal in .NET 6). This issue is pretty insidious because the example in the description shows a broken case, but the overwhelming majority of cases work, which is why I have a problem with the PR in the first place.

We have a gap with what we can express currently and a better plan is to use the analyzers/linker knowledge to fill it. I don't like the linker specific solution because there are other libraries that will have the same problem I wouldn't recommend they do what this PR does, it's just not right.

MichalStrehovsky commented:
Fully trimmable DependencyInjection is tracked as a Pri2 issue here: https://github.com/dotnet/runtime/issues/44432. The solution would have been a source generator but it wasn't done because it's a Pri2.

davidfowl commented:
That issue as it currently stands is a pipe dream 😄. The only gap that I'm aware of right now (which this PR is trying to address) is open generics that need to express dynamic behavior over their generic arguments when the service type and implementation type are different. That very specific scenario should warn ONLY if the service type (typically the interface) and implementation type have differing DAM attributes.

davidfowl commented:
To make that concrete. This would print nothing when trimmed because of the mismatch in DAM attributes between `IMemberPrinter<T>` and `MemberPrinter<T>`

```C#
using Microsoft.Extensions.DependencyInjection;
using System;
using System.Diagnostics.CodeAnalysis;

IServiceProvider sp = new ServiceCollection()
            .AddSingleton(typeof(IMemberPrinter<>), typeof(MemberPrinter<>))
            .BuildServiceProvider();

var factory = sp.GetService<IMemberPrinter<Customer>>();

factory.PrintMembers();
Console.Read();

interface IMemberPrinter<T>
{
    void PrintMembers();
}

class MemberPrinter<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties)]T> : IMemberPrinter<T>
{
    public void PrintMembers()
    {
        foreach (var p in typeof(T).GetProperties())
        {
            Console.WriteLine(p.Name);
        }
    }
}

class Customer
{
    public int Id { get; set; }
    public string Name { get; set; }
}
```

eerhardt commented:
What's the path forward here? I don't believe the current experience is "good enough" to ship in 6. I think we should do something to address this. Options forward that I see are:

1. Take the approach used in this PR. Make it noisy, and force most people to suppress the warning.
2. Add a rule into mono/linker (and I guess NativeAOT, if necessary) that finds callers to the 20 APIs marked here, checks if the arguments are open generic types, and if they are, ensures the `[DAM]` attributes match between the two.
3. Change the DependencyInjection code and leave in a "LibraryBuild" warning that gives a single, global warning whenever DependencyInjection is used. The warning has a good message saying "Using generic types in DependencyInjection can lead to required code being trimmed. Ensure any generic Service types ...". Doing this means that developers will see the warning when trimming, instead of getting the collapsed warning experience we have today. (see https://github.com/mono/linker/issues/2004 and https://github.com/mono/linker/pull/2087)
4. Add a separate Roslyn analyzer if we aren't comfortable with adding the rule to mono/linker.
5. Add a new "generalized feature" in mono/linker that allows for recognizing this pattern more generally
    - I'm not sure how this would work. Maybe an attribute that could be applied on one parameter that points to another parameter and says "the [DAM] attributes need to match between these two parameters"
6. Others?


________

A trap with the current approach taken with this PR is that it will easily lead to people not seeing the warnings. Say in "v1" of my app, I have a ConfigureServices method:

```C#
public static void ConfigureServices(IServiceCollection services)
{
    services.AddSingleton<IMyService, MyService>();
}
```

I get a warning saying "Using generic types in DependencyInjection can lead to required code being trimmed.". I decide "I'm not using generic types here. Let me add a suppression:

```diff
+[UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode"]
public static void ConfigureServices(IServiceCollection services)
{
    services.AddSingleton<IMyService, MyService>();
}
```

Now in "v2" of my app, I add a new feature:

```diff
[UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode"]
public static void ConfigureServices(IServiceCollection services)
{
    services.AddSingleton<IMyService, MyService>();
+   services.AddSingleton<IMyGenericService<>, MyGenericService<>>();
}
```

Everything is fine, no warnings, because in `v1` I suppressed a noisy warning. But now in `v2` my app is broke because the existing suppression hid the new, valid warning.

jkotas commented:
I like option 4.

jkotas commented:
FWIW, I think punting this for .NET 6 would be fine too. I do not think that this moves the needle enough for trimming warnings in appmodels that use DI.

davidfowl commented:
I vote do nothing in .NET 6 but document the behavior and show the workaround. Then if we have time we do the analyzer out of band

eerhardt commented:
After discussing more with @vitek-karas, we came up with a few more thoughts:

### Not just a trimming problem
The issue with how open generics works in DependencyInjection isn't just a trimming issue. You get a runtime error with the following app:

```C#
using Microsoft.Extensions.DependencyInjection;
using System;

IServiceProvider sp = new ServiceCollection()
            .AddSingleton(typeof(INullablePrinter<>), typeof(NullablePrinter<>))
            .BuildServiceProvider();

var printer = sp.GetService<INullablePrinter<Customer>>();
printer.PrintNullable();

interface INullablePrinter<T>
{
    void PrintNullable();
}

class NullablePrinter<T> : INullablePrinter<T> where T : struct  // ISSUE HERE - generic constraints don't match with INullablePrinter<T>
{
    public void PrintNullable()
    {
        Nullable<T> n = default;
        Console.WriteLine(n);
    }
}

class Customer { }
```

```
Unhandled exception. System.ArgumentException: GenericArguments[0], 'Customer', on 'NullablePrinter`1[T]' violates the constraint of type 'T'.
```

### Option 4 isn't enough by itself

Having an analyzer that isn't the mono/linker isn't enough by itself. We still need to ensure there are warnings in the app whether the analyzer runs or not. Either:
1. the current warning we have today
2. option 1 above
3. or option 3 above

### More Potential Options
6. Add a runtime check to DependencyInjection that would check open generic types have matching `[DAM]` attributes. This check would happen regardless if the app was trimmed or not. The drawback here would be that if the app was only ever run as a trimmed app, the `[DAM]` attributes could be trimmed, and the runtime check wouldn't catch the issue. So we'd still need at least some warning here - same as Option 4.

eerhardt commented:
> I vote do nothing in .NET 6 but document the behavior and show the workaround.

I'd at least do Option 3, so the warning that devs get can point to the document, and they can semi-understand what the problem is.

MichalStrehovsky commented:
Could DI validate the DAM annotations on whatever is passed to it and ensure they're consistent before it feeds things to MakeGenericType? It violates separation of concerns but at least it would ensure the same behavior before/after trimming. DI would be suppressing the warning, so that's why it should be on the hook for validation. Maybe it would be useful to have a reflection-API-based validator for this for the future too.

Of course we drop DAM attributes after trimming so this would not validate anything after trimming is done, but we're after crippling the F5 behavior so that it's not more permissive than once we're done with trimming.

davidfowl commented:
I love this reframing of the problem. It's a general constraints problem and the generic implementation type and generic service type needs to have compatible constraints or there will be a runtime failure. I think we could add a runtime check when enumerating open generic service descriptors that checks DAM compatibility explicitly. We already have a couple of checks for open generics, this would be no different:

https://github.com/dotnet/runtime/blob/bb3982239e11d559f839aba445ea7d698084e623/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs#L37-L59

vitek-karas commented:
I like the idea from @MichalStrehovsky to "break" the bad case in the runtime always. It would at least make DAM and generic constraints behave the same. The question here is if we can do it. It's technically a breaking change. Apps which don't care about trimming would still be affected by DAM annotations. Combined with us adding DAM annotations in the framework, this could lead to some "not-very-nice" breaks, which are in reality a non-issue (even if DAM doesn't match, as long as the app is not trimmed, things will just work fine - unlike the generic constains).

MichalStrehovsky commented:
> Combined with us adding DAM annotations in the framework, this could lead to some "not-very-nice" breaks, which are in reality a non-issue (even if DAM doesn't match, as long as the app is not trimmed, things will just work fine - unlike the generic constains).

We could set an AppContext switch when PublishTrimmed is specified in the CSPROJ and limit the enforcement to when the AppContext switch is set. It might be useful to have such documented thing available in general.

davidfowl commented:
This comes down to the thing we spoke about before where it should be possible to tell at runtime if trimming took place.

MichalStrehovsky commented:
> This comes down to the thing we spoke about before where it should be possible to tell at runtime if trimming took place.

The AppContext switch would be set irrespective of whether the app is trimmed. It would be set if the app can potentially be trimmed (i.e. it would also be set when F5 debugging an app that is opted into publish-time trimming through a CSPROJ property). We don't want different behavior between F5 debug and trimmed publish.

davidfowl commented:
> The AppContext switch would be set irrespective of whether the app is trimmed. It would be set if the app can potentially be trimmed (i.e. it would also be set when F5 debugging an app that is opted into publish-time trimming through a CSPROJ property). We don't want different behavior between F5 debug and trimmed publish.

OK this sounds like a good idea. Do we want a more first class API as well?

MichalStrehovsky commented:
> OK this sounds like a good idea. Do we want a more first class API as well?

I don't have an opinion about that. It feels pretty niche and this might be the first spot where we have a need for it. Maybe no API for now and we can re-evaluate when we have more data?

davidfowl commented:
Sounds good

davidfowl commented:
@eerhardt do we file a follow up item and close this PR?

eerhardt commented:
I think I will just modify this PR with the plan we decide on.

My understanding of the plan, please correct me if I'm wrong:

1. We leave a "LibraryBuild" warning in DependencyInjection that has a good warning message explaining that open generic types can be broken. Maybe even putting an `aka.ms` link to a doc explaining the situation.
2. We add a new runtime check to DependencyInjection that verifies when open generic types are used, they have matching `[DAM]` annotations to the generic types.
    a. The check is only enabled when a new AppContext switch (`Microsoft.Extensions.DependencyInjection.VerifyOpenGenericServiceTrimmability`) is set. By default it isn't set.
3. We enable `VerifyOpenGenericServiceTrimmability` when `PublishTrimmed=true` on the project.

We are recommending that app devs set `PublishTrimmed=true` in their .csproj when they plan on trimming their apps. That way feature switches like BuiltInCOMInterop are configured correctly even when F5 Debugging their untrimmed apps. And the analyzer runs by during the build. `VerifyOpenGenericServiceTrimmability` is just another setting that gets defaulted with PublishTrimmed. That way devs will see any errors even before trimming their apps.

MichalStrehovsky commented:
> We leave a "LibraryBuild" warning in DependencyInjection that has a good warning message explaining that open generic types can be broken. Maybe even putting an aka.ms link to a doc explaining the situation.

Under what circumstances will the warning show up?

The rest of the plan looks good to me!

davidfowl commented:
I think it always shows up. 

> Microsoft.Extensions.DependencyInjection.VerifyOpenGenericServiceTrimmability

Don't we have a more generic name for `<PublishTrimmed>` is enabled?

eerhardt commented:
>>Under what circumstances will the warning show up?

> I think it always shows up.

Correct. If your app uses Microsoft.Extensions.DependencyInjection, the warning will show up.

>> Microsoft.Extensions.DependencyInjection.VerifyOpenGenericServiceTrimmability

> Don't we have a more generic name for `<PublishTrimmed>` is enabled?

We don't have a runtime switch that checks "is this app going to be trimmed?" Each library defines their switches as necessary, and then those switches can be defaulted accordingly. That way, no matter what `PublishTrimmed` is set to, the switch can be set to the developer's desired behavior.

MichalStrehovsky commented:
> Correct. If your app uses Microsoft.Extensions.DependencyInjection, the warning will show up.

Why do we want to warn if the behavior before/after trimming is going to be consistent?

One can get behavioral differences if one goes and turns VerifyOpenGenericServiceTrimmability off manually after setting PublishTrimmed, but we don't care about people doing that. If someone goes and turns COM/StartupHooks/etc. back on, they took off the safety guards and they're on their own.

eerhardt commented:
> Why do we want to warn if the behavior before/after trimming is going to be consistent?

The behavior isn't going to be consistent before/after trimming. Someone could run their app all day without trimming, and it works. Then they `dotnet publish -p:PublishTrimmed=true`, and now the app is broken (because the trimmer removed the `[DAM]` attributes - no runtime error occurs). There is no guarantee they added `PublishTrimmed` to their .csproj. We recommend it, but how do they know about that recommendation?

MichalStrehovsky commented:
> There is no guarantee they added PublishTrimmed to their .csproj. We recommend it, but how do they know about that recommendation?

The same problem exists for startup hooks or COM. The "trimming emulator" depends on PublishTrimmed being set in CSPROJ and not being specified as a publish option. There's nothing better we can do about it. When PublishTrimmed is specificed, we pull the plug on startup hook or COM. If the user was F5 debugging without PublishTrimmed being set in their csproj and they did `dotnet publish -p:PublishTrimmed=true` instead, now they have a broken app.

See https://github.com/mono/linker/issues/2085

eerhardt commented:
> The same problem exists for startup hooks or COM.

Yep, and we warn in both of those cases as well.

davidfowl commented:
What do you do about warnings that never go away? You suppress them 😄

MichalStrehovsky commented:
> Yep, and we warn in both of those cases as well.

We don't. PublishTrimmed will disable all of StartupHook through a feature switch and there will be no build time warning that it stopped working. Otherwise a hello world would need to warn.

eerhardt commented:
> We don't.

When the potentially unsafe behavior is enabled, we warn. StartupHooks are disabled by default, so we don't warn. When you enable it, we warn.

We aren't disabling the unsafe behavior in DI, so we warn.

davidfowl commented:
When a warning is always on, it becomes useless. If every ASP.NET Core application always has this warning then its pretty useless.

agocke commented:
My understanding is that this warning is pointing out a real incompatibility with trimming. Is that right?

If so, I think the warning is appropriate. There are other libraries, like System.Xml, which are not trimming compatible, and the only recourse is to remove the reference from the application.

davidfowl commented:
While technically correct, it doesn't make this warning anymore useful for actionable and as a result, it will be disabled.

eerhardt commented:
Agreed - people will disable this warning, a lot, or just ignore it. But hopefully they will become educated about the issue and understand the problem. Then they know how to fix their code, so their apps don't break after trimming.

eerhardt commented:
Note - I still believe the best user experience option would be:

2. Add a rule into mono/linker (and I guess NativeAOT, if necessary) that finds callers to the 20 APIs marked here, checks if the arguments are open generic types, and if they are, ensures the `[DAM]` attributes match between the two.

This wouldn't give "false-positive" warnings that need to be suppressed. The rule would only warn when the developer actually had an issue with their code.

MichalStrehovsky commented:
We need to warn whenever the app could behave differently after trimming. There are two ways how that can happen:

1. User sets PublishTrimmed as part of publish and not in csproj. This is not a scenario we optimize for. If it was, hello world needs to warn because startup hook behavior could change.
2. User opts out of VerifyOpenGenericServiceTrimmability. This is a potential concern, albeit minor.

2 can by solved by doing roughly this:

```
[UnconditionalSuppress]
Void ValidateAndCreate() {}

Void CreateButDontValidate() {}

If (Feature.Validate)
    ValidateAndCreate()
Else
    CreateButDontValidate()
```

Illink will eliminate the unsuppressed code when the feature is opted in, but issue a warning (because there is no suppression) if the feature is not enabled.

Does that address the concern or am I missing something?

eerhardt commented:
I think the part you may be missing is that the `[DAM]` attributes get removed during trimming.

https://github.com/dotnet/runtime/blob/e0024e22c7761cbbdd10a845f8b0269ea654cbfc/src/mono/System.Private.CoreLib/src/ILLink/ILLink.LinkAttributes.xml#L118-L120

This means:

1. Write the app in the top comment of this post, but without `<PublishTrimmed>true</PublishTrimmed>` in the .csproj
2. Publish and run the app.
3. Now add `<PublishTrimmed>true</PublishTrimmed>` to the .csproj
4. Publish and run the app.

The app behavior changes. In step (2) the app worked. In step (4) the app throws `Unhandled exception. System.MissingMethodException: Cannot dynamically create an instance of type 'ConsoleApp1.Customer'. Reason: No parameterless constructor defined.`.

IMO, that behavior violates your above statement:

> We need to warn whenever the app could behave differently after trimming.

eerhardt committed:
Resolve MakeGenericType ILLink warning in DependencyInjection

Resolve the ILLink warning in DependencyInjection by adding a runtime check that is behind a new AppContext switch 'Microsoft.Extensions.DependencyInjection.VerifyOpenGenericServiceTrimmability'. The runtime check ensures the trimming annotations on the open generic types are compatible between the service and implementation types. The check is enabled by default when PublishTrimmed=true.

eerhardt commented:
I've updated this PR with the [above plan](https://github.com/dotnet/runtime/pull/55102#issuecomment-881168096) (excluding leaving a LibraryBuild warning in). We believe the runtime check that is enabled when PublishTrimmed=true will be enough notification to the developer.

Please take a look at the new approach.

davidfowl commented:
> (excluding leaving a LibraryBuild warning in)

😢

eerhardt commented:
>> (excluding leaving a LibraryBuild warning in)

>😢

I thought you [didn't want the warning](https://github.com/dotnet/runtime/pull/55102#issuecomment-882683595).

davidfowl commented:
Oh wait, I don't 😄

vitek-karas commented:
Sounds good - if we figure out a way to check for this statically we can add a new warning in future releases.

eerhardt committed:
Make VerifyOpenGenericServiceTrimmability a full feature switch

agocke commented:
I also noticed that there are now multiple source-generator-based dependency injection libraries, e.g.

https://github.com/giggio/sourceinject/
https://github.com/pakrym/jab

which promise AOT-friendly codegen for dependency injection. These libraries could be alternatives for users if we find more problems than expected.

jkotas commented:
Yes, there will be more problems. The current pattern, even with this change, is not AOT friendly. Do we have a solution in mind that would make it AOT friendly?

> I also noticed that there are now multiple source-generator-based dependency injection libraries

Yep, we will know that trimming and AOT friendliness is taking off once there is something like https://quarkus.io/ for .NET.

eerhardt commented:
> The current pattern, even with this change, is not AOT friendly. Do we have a solution in mind that would make it AOT friendly?

AFAIK, no. If we limited the generic parameter types to "ref" types, would that make it AOT friendly? If yes, maybe that's just what we instruct library and app developers who want to use this feature with AOT.

The open generics feature is used in a few pretty core places:

* Options
https://github.com/dotnet/runtime/blob/cf52b7e38d28b7e28358cee052effe0ed47efb2e/src/libraries/Microsoft.Extensions.Options/src/OptionsServiceCollectionExtensions.cs#L29-L33
* Http
https://github.com/dotnet/runtime/blob/f5a73049046e6267fbfcc5d8ebbf3c88d8cdea1e/src/libraries/Microsoft.Extensions.Http/src/DependencyInjection/HttpClientFactoryServiceCollectionExtensions.cs#L44-L45

davidfowl commented:
> Yes, there will be more problems. The current pattern, even with this change, is not AOT friendly. Do we have a solution in mind that would make it AOT friendly?

What isn't AOT friendly?

agocke commented:
I believe all uses of `MakeGenericType` are fundamentally AOT unfriendly as they require runtime code generation.

davidfowl commented:
> I believe all uses of MakeGenericType with structs are fundamentally AOT unfriendly as they require runtime code generation.

I don't see why if these are reference types.

agocke commented:
Is there any protection against registering structs?

davidfowl commented:
For arbitrary open generics no.

davidfowl commented:
I think the fear that this pattern isn't AOT friendly is a "perfect is the enemy of good" situation. I think this is pretty good already. I can see a future where we have a new method designed for open generics that would allow us to add more static analysis where possible.

eerhardt commented:
Any feedback here? I believe this is ready to merge.

eerhardt committed:
Merge branch 'main' into DIILink

eerhardt committed:
Merge branch 'main' into DIILink

eerhardt committed:
Fix up merge
